text/0000-pub-sub.md

```diff
@@ -0,0 +1,171 @@
+  - `DELETE /pubsub/eventtypes/declare/publisher` - to remove publisher declaration
+  - `POST /pubsub/eventtypes/declare/subscriber` - to declare a subscriber for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/subscriber` - to remove subscriber declaration
+- API to retrieve existing registered event types.
+  - `GET /pubsub/eventtypes` - returns all registered event types
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}` - returns an event descriptor for a particular event type
+- API to retrieve the registered subscriptions for a specified tenant and event type
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}/subscribers` - returns all subscribers of the event type provided in the path variable
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}/publishers` - returns all publishers of the event type provided in the path variable
+- API used by publishers to send events. This API must be called by back-end modules which act as publishers
+  - `POST /pubsub/publish`
+- API to retrieve the activity history performed by the event manager. A correlationID can optionally be provided to limit context.
+  - `GET /pubsub/history`
+  - `GET /pubsub/history?corellationId={corellationId}`
+
+It is allowed for publishers to provide callback URLs to receive feedback from `mod-pubsub` module in cases when:
+- Subscriber did not receive an event within the specified period of time.
+- All subscribers deleted their subscriptions on a particular event type registered by this publisher.
+ 
+
+There is a set of requirements that must be followed by a module to provide “Subscriber endpoints” to consume events. The basic set of requirements:
+- HTTP Method - POST
+- A set of required HTTP header. [_To be defined further if it is needed._]
+- A payload structure - JSON which represents an event. See [Event data structure](#Event data structure) 
+
+Reference implementations should be provided that allow a back-end module to become a publisher or a subscriber respectively to simplify development.
+  
+A dedicated module, `mod-pubsub`, must implement and encapsulate all logic related to event sending and delivery. All interactions between this module and back-end business modules will be done via HTTP.
+
+#### Registration
+
+Each module which acts as a publisher or a subscriber must contain a configuration or property with a set of event types it deals with. Publishers should provide event descriptors. For subscribers, it is enough to specify event types together with endpoints for event delivery. Details of that configuration must be defined further.
+
+##### Event Type registration 
+The set of event descriptors will be registered by a publisher at the time when the module is enabled for a tenant. To do that the publisher must call `pubsub` API for event type registration. The payload for that call is a set of event descriptors which define events this particular module is able to publish.
```

Comment:
Is the endpoint in `mod-pubsub` for event type registration likely to required permissions?

If so, any module that registers event types will need that permission as a module permission for the Tenant API activation endpoint.

Reply:
What should happen if that registration fails? Should tenant activation for that module fail?

Reply:
Yes and yes. Specifically if the registration is happening during tenant activation. 

However, since this is a dynamic process, it is possible that some event subscriptions may not be declared at tenant activation time. In which case, the subscription declaration might be done at a later time and it would be inappropriate to deactivate the tenant's module on the failed declaration.

Reply:
Thanks for clarifying that.

Sure, dynamic registration after activation is different